### PR TITLE
enables listing apps by individual service description parameters

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- enables listing apps by individual service description parameters

## Why are we making these changes?

Simplifies searching for services by their service description parameters.

### CODE REVIEW

File: waiter/src/waiter/handler.clj [hunk 1]
```diff
@@ -254,34 +254,57 @@
     :scale-amount
     utils/scale-amount->scaling-state))
 
+(defn param-value->filter-fn
+  "Creates the filter function that does substring match for the input string or any string in the sequence."
```
Comment (scrosby) [inline waiter/src/waiter/handler.clj]: This description is unclear. Maybe something like:

"Takes a single string or a sequence of strings. Returns a filter function ...."

But then, that description makes this smell like it should be split into two functions  strs->filter-fn .... which handles the multiple case (the second branch of the if) and then this functon. Or maybe after splitting, inline this function into its caller?

File: waiter/src/waiter/handler.clj [hunk 2]
```diff
@@ -254,34 +254,57 @@
...
-                                  (and (if (str/blank? run-as-user)
-                                         (authz/manage-service? entitlement-manager auth-user service-id service-description)
-                                         ((str->filter-fn run-as-user) (get service-description "run-as-user")))
-                                       (or (str/blank? token)
-                                           (let [filter-fn (str->filter-fn token)]
-                                             (->> source-tokens
+          viewable-service-ids (filter
+                                 (fn [service-id]
+                                   (let [service-description (service-id->service-description-fn service-id :effective? true)
+                                         source-tokens (-> (service-id->source-tokens-entries-fn service-id) vec flatten)]
```
Comment (scrosby) [inline waiter/src/waiter/handler.clj]: Why merge this into the same line?

File: waiter/src/waiter/handler.clj [hunk 3]
```diff
@@ -254,34 +254,57 @@
...
+                                     (and (service-description-filter-predicate service-description)
+                                          (if (or (nil? run-as-user)
```
Comment (scrosby) [inline waiter/src/waiter/handler.clj]: I assume this nil? guard was to protect against a bug? Why only here, not token-version? or token?

File: waiter/src/waiter/handler.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: Created `strs->filter-fn`

File: waiter/src/waiter/handler.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: Minor change as I find it easier to read having to parse fewer lines.

File: waiter/src/waiter/handler.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: The `nil` behavior for `run-as-user` is different for legacy reasons. When no `run-as-user` is provided, we display only services a user can manage.

**Pull request merged**